lib/pg_ha_migrations/safe_statements.rb

```diff
@@ -280,10 +288,171 @@ def safe_create_partitioned_table(table, key:, type:, infer_primary_key: nil, **
         arr << col.name
       end
 
-      if infer_primary_key && !key.is_a?(Proc) && ActiveRecord::Base.connection.postgresql_version >= 11_00_00
-        td.primary_keys(pk_columns.concat(Array.wrap(key)).map(&:to_s).uniq)
+      if infer_primary_key && !partition_key.is_a?(Proc) && ActiveRecord::Base.connection.postgresql_version >= 11_00_00
+        td.primary_keys(pk_columns.concat(Array.wrap(partition_key)).map(&:to_s).uniq)
+      end
+    end
+  end
+
+  def safe_partman_create_parent(table, **options)
+    if options[:retention].present? || options[:retention_keep_table] == false
+      raise PgHaMigrations::UnsafeMigrationError.new(":retention and/or :retention_keep_table => false can potentially result in data loss if misconfigured. Please use unsafe_partman_create_parent if you want to set these options")
+    end
+
+    unsafe_partman_create_parent(table, **options)
+  end
+
+  def unsafe_partman_create_parent(
+    table,
+    partition_key:,
+    interval:,
+    infinite_time_partitions: true,
+    inherit_privileges: true,
+    premake: nil,
+    start_partition: nil,
+    template_table: nil,
+    retention: nil,
+    retention_keep_table: nil
+  )
+    raise ArgumentError, "Expected <partition_key> to be present" unless partition_key.present?
+    raise ArgumentError, "Expected <interval> to be present" unless interval.present?
+
+    if ActiveRecord::Base.connection.postgresql_version < 11_00_00
+      raise PgHaMigrations::InvalidMigrationError, "Native partitioning with partman not supported on Postgres databases before version 11"
+    end
+
+    formatted_start_partition = nil
+
+    if start_partition.present?
+      # Rails 7+ prefers to_fs(:db) and has deprecated to_s(:db)
+      %i[to_fs to_s].each do |meth|
+        begin
+          formatted_start_partition = start_partition.send(meth, :db)
+        rescue NoMethodError, ArgumentError
+        end
+
+        break if formatted_start_partition.present?
+      end
+
+      if !formatted_start_partition.present?
+        raise PgHaMigrations::InvalidMigrationError, "Expected <start_partition> to be Date, Time, or DateTime object but received #{start_partition.class}"
```

Comment:
The goal of this `if` block was to allow duck-typed date / time objects while also using the appropriate method (`to_fs` vs `to_s`) for different rails versions, but I don't really love this code.

It would probably be more readable if we just check if `start_partition.is_a?` date / time object upfront and then call the appropriate method with a conditional (`if rails < 7 do this else do that`)

Reply:
Yeah I was trying to be way too clever here. I rewrote this and pushed a change